README
Tk-FileEntry
============

This Perl/Tk widget provides an Entry widget with a button. Clicking
the button will open a choose file dialog. The path of the chosen file
will be inserted into the Entry widget.

This is useful if you want to provide a convenient way to select a
file path. Otherwise, you would have to code this feature over and
over again.

You can read a nicely formatted version of the documentation for
this module online:

    https://metacpan.org/pod/Tk::FileEntry


INSTALLATION

You should be able to install this using your usual method for installing
modules from CPAN. If you don't have one, have a look at:

    http://www.cpan.org/modules/INSTALL.html


DEPENDENCIES

This module requires some other modules and libraries.
See Makefile.PL for a detailed list. If you are going to install
via cpan shell or PPM, you don't have to care about.


COPYRIGHT AND LICENSE

Copyright (C) 2013-2020 by Alexander Becker

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.16.3 or,
at your option, any later version of Perl 5 you may have available.

Changes
Revision history for Perl extension Tk::FileEntry.

2.4  2020-10-28
    - do not test POD during install (check for RELEASE_TESTING environment variable)
    - cosmetic changes: fix spelling issues, author reference, indentation

2.3  2015-03-08
    - added LICENSE file
    - reworked readme

2.2  2014-08-12
    - lowered Perl requirement to 5.8.8, related to RT #97845 by Slaven

2.1  2014-07-30
    - fixed timestamp of changes entry for v2.0
    - skipped test using MainWindow for computers not having a display
    - removed self-made test suite and replaced by Test::More
    - bumped Tk requirement to Tk 804.030 because I can't / don't want to test for an older Tk
    - removed bugs section from POD because bumping Tk version requirement makes bug notice obsolete
    - removed METHDOS section from POD as there are no met hods to document
    - removed duplicate BUGS section
    - added example to POD

2.0  2014-07-29
    - pimped original version by moving it to the skeleton created by h2xs 1.23 with options -b 5.14.3 -XA -n Tk::FileEntry
    - added Changes file
    - removed revision control version number stuff
    - removed print command for arguments
    - user base instead of direct use of @ISA
    - moved POD inside the code, documenting all methods
    - extended documentation
    - switched from Tk::FileSelect to getOpenFile
    - removed requires and added it with use pragma (if you are going to use this module, then you are going to use the dependencies as well)
    - removed known bug POD related to FileSelect issue - it's not used anymore
    - extended fileentry.t

1.4  2013-11-26
    - version found on CPAN in Tk::ACH dist v0.04 and on gitpan

Makefile.PL
use 5.008008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Tk::FileEntry',
    VERSION_FROM      => 'lib/Tk/FileEntry.pm', # finds $VERSION
    ABSTRACT_FROM     => 'lib/Tk/FileEntry.pm', # retrieve abstract from module
    PREREQ_PM         => {
        # e.g., Module::Name => 1.1
        'ExtUtils::MakeMaker' => '6.64',
        'Tk' => '804.030',
    },
    TEST_REQUIRES => {
        # modules only required for testing
        'Test::More' => '0.98',
        'Test::Pod' => '1.0',
        'Test::Pod::Coverage' => '1.08',
    },
    AUTHOR         => 'Alex Becker <asb@cpan.org>',
    LICENSE        => 'perl_5',
    dist           => { COMPRESS => "gzip", SUFFIX => "gz" },
    META_MERGE     => {
        resources => {
            repository => 'https://github.com/asb-capfan/Tk-FileEntry',
        },
    },
    MIN_PERL_VERSION => 5.008008,
);


lib/Tk/FileEntry.pm
package Tk::FileEntry;

use 5.008008;
use strict;
use warnings;
use Tk;
use Tk::widgets qw/ Frame Derived Widget Label Entry Button /;
use base qw/ Tk::Derived Tk::Frame /;

our $VERSION = '2.4';

Construct Tk::Widget 'FileEntry';

my $FILEBITMAP = undef;

=head1 NAME

Tk::FileEntry - FileEntry widget with optional file selection box

=head1 SYNOPSIS

    use Tk::FileEntry;

    $fileentry = $parent->FileEntry(
        -filebitmap	=> BITMAP,
        -command	=> CALLBACK,
        -variable	=> SCALARREF,
	);

=head1 DESCRIPTION

FileEntry is a composite widget for choosing files.
It features a L<Tk::Label>, L<Tk::Entry>, and a L<Tk::Button>. 

When the button is clicked, a dialog for choosing a file will show up.
The path of the chosen file will be inserted into the entry widget.
The label is intended as caption for the entry widget.

This is useful if you want to provide a convenient way to select a
file path.




=head1 WIDGET-SPECIFIC OPTIONS

=over 4

=item C<Option:> B<-filebitmap>

=item C<Name:> B<fileBitmap>

=item C<Class:> B<FileBitmap>

Specifies the bitmap to be used for the button that invokes the File Dialog.

=item B<-command>

A callback that is executed when a file is chosen.

Pressing enter in the entry widget will execute this callback. See L<Tk::FileEntry/BINDINGS>.

=item B<-variable>

Reference to variable that will be bound to the value of the entry widget.
See C<Tk::options> for more details on C<-variable>.

=item B<-label>

Defines the label text. Defaults to I<File:>.

=back

=cut


# METHODS

sub ClassInit {
    my ($class, $mw) = @_;

    return if defined $FILEBITMAP;  # needed for several MainWindows
    $FILEBITMAP = __PACKAGE__ . '::OPENFOLDER';

    my $bits = pack("b16"x10,
        "...111111.......",
        "..1......11.....",
        ".1.........1....",
        ".1..........1...",
        ".1...11111111111",
        ".1..1.1.1.1.1.1.",
        ".1.1.1.1.1.1.1..",
        ".11.1.1.1.1.1...",
        ".1.1.1.1.1.1....",
        ".1111111111.....",
        );

    $mw->DefineBitmap($FILEBITMAP => 16,10, $bits);

}

sub Populate {
    my ($w,$args) = @_;

    $w->SUPER::Populate($args);

    my $l = $w->Label()->pack(-side=>'left');
    my $e = $w->Entry()->pack(-side=>'left', -expand=>'yes', -fill=>'x');
    my $b = $w->Button(
		-command => [\&_selectfile, $w, $e],
		-takefocus => 0,
	)->pack(
		-side => 'left',
		-fill => 'y',
	);

    $e->bind('<Return>', [$w, '_invoke_command', $e]);

    $w->Advertise('entry' => $e);
    $w->Advertise('button' => $b);

	$w->Delegates(
		'get' => $e,
		'insert' => $e,
		'delete' => $e,
		DEFAULT => $w,
	);
	
    $w->ConfigSpecs(
	    -background	=> [qw(CHILDREN background Background), Tk::NORMAL_BG()],
	    -foreground	=> [qw(CHILDREN foreground Foreground), Tk::BLACK()    ],
	    -state 	=> [qw(CHILDREN state      State        normal)        ],
	    -label       => [{-text => $l},   'label',      'Label',     'File:'],
	    -filebitmap  => [{-bitmap => $b}, 'fileBitmap', 'FileBitmap', $FILEBITMAP,],
	    -command	 => ['CALLBACK',       undef,        undef,       undef],
	    -variable    => ['METHOD',         undef,        undef,       undef],
	);
	
    return $w;
} # /populate




sub _selectfile {
    my $w = shift;
    my $e = shift;

	my $file = $w->getOpenFile();

    return unless defined $file && length $file;
    $e->delete(0,'end');
    $e->insert('end',$file);
    $w->Callback(-command => $w, $file);
}


sub _invoke_command {
    my $w = shift;
    my $e = shift;
    my $file = $e->get();
    return unless defined $file && length $file;
    $w->Callback(-command => $w, $e->get);
}


# variable( $v ) is listed as method in ConfigSpecs

sub variable {
    my $e = shift->Subwidget('entry');
    my $v = shift;
    $e->configure(-textvariable => $v);
}    




=head1 BINDINGS

C<Tk::FileEntry> has default bindings to allow the execution of the callback when a user presses enter in the entry widget.




=head1 EXAMPLE

  use strict;
  use warnings;
  use Tk;
  use Tk::FileEntry;
  
  my $mw = tkinit();
  
  $mw->FileEntry->pack(-expand => 1, -fill => 'x');
  
  $mw->MainLoop();



=head1 BUGS

None yet. If you find one, please consider creating a bug report, e.g. via L<Github|https://github.com/asb-capfan/Tk-FileEntry/issues>.




=head1 SEE ALSO

=over

=item * L<Tk::getOpenFile> for details about the file selection dialog

=item * L<Tk::Entry> for details about the Entry widget

=item * L<Tk> for details about the Perl/Tk GUI library

=back

There is a wiki for Tcl/Tk stuff on the web: L<http://wiki.tcl.tk/>




=head1 KEYWORDS

fileentry, tix, widget, file selector




=head1 AUTHOR

Alexander Becker, E<lt>asb@cpan.orgE<gt>

Original Author was Achim Bohnet E<lt>ach@mpe.mpg.deE<gt>.

This code is inspired by the documentation of FileEntry.n of the Tix4.1.0 distribution by Ioi Lam.
The bitmap data are also from Tix4.1.0. For everything else:

=head1 COPYRIGHT

Copyright (C) 2013-2020 by Alexander Becker

Copyright (c) 1997-1998 Achim Bohnet. All rights reserved.

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.16.3 or,
at your option, any later version of Perl 5 you may have available.


=cut


1; # /Tk::FileEntry